Source/Core/VideoCommon/OnScreenUI.cpp

```diff
@@ -340,43 +340,54 @@ void OnScreenUI::DrawChallengesAndLeaderboards()
+      auto res = m_challenge_texture_map.insert_or_assign(name, g_gfx->CreateTexture(tex_config));
+      res.first->second->Load(0, width, height, width, icon->data.data(),
+                              sizeof(u32) * width * height);
+    }
+    float window_width = 0;
+    float window_height = 0;
+    for (auto& [name, texture] : m_challenge_texture_map)
+    {
+      auto icon_itr = challenge_icons.find(name);
+      if (icon_itr == challenge_icons.end())
+      {
+        m_challenge_texture_map.erase(name);
+        continue;
+      }
+      window_width += icon_itr->second->width;
+      window_height = std::max(window_height, static_cast<float>(icon_itr->second->height));
+    }
+    window_width += ImGui::GetStyle().WindowPadding.x * 2;
+    window_height += ImGui::GetStyle().WindowPadding.y * 2;
+    ImGui::SetNextWindowSize(ImVec2(window_width, window_height));
+    ImGui::SetNextWindowPos(ImVec2(ImGui::GetIO().DisplaySize.x - window_width,
+                                   ImGui::GetIO().DisplaySize.y - window_height),
+                            0);
     if (ImGui::Begin("Challenges", nullptr,
                      ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoInputs |
                          ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoSavedSettings |
                          ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoNav |
                          ImGuiWindowFlags_AlwaysAutoResize | ImGuiWindowFlags_NoFocusOnAppearing))
     {
-      for (const auto& [name, icon] : challenge_icons)
-      {
-        if (m_challenge_texture_map.find(name) != m_challenge_texture_map.end())
-          continue;
-        const u32 width = icon->width;
-        const u32 height = icon->height;
-        TextureConfig tex_config(width, height, 1, 1, 1, AbstractTextureFormat::RGBA8, 0,
-                                 AbstractTextureType::Texture_2DArray);
-        auto res = m_challenge_texture_map.insert_or_assign(name, g_gfx->CreateTexture(tex_config));
-        res.first->second->Load(0, width, height, width, icon->data.data(),
-                                sizeof(u32) * width * height);
-      }
+      ImVec2 cursor_position = ImGui::GetCursorPos();
       for (auto& [name, texture] : m_challenge_texture_map)
       {
         auto icon_itr = challenge_icons.find(name);
-        if (icon_itr == challenge_icons.end())
-        {
-          m_challenge_texture_map.erase(name);
-          continue;
-        }
-        if (texture)
-        {
-          ImGui::Image(texture.get(), ImVec2(static_cast<float>(icon_itr->second->width),
-                                             static_cast<float>(icon_itr->second->height)));
-        }
+        ImGui::SetCursorPos(cursor_position);
+        ImGui::Image(texture.get(), ImVec2(static_cast<float>(icon_itr->second->width),
+                                           static_cast<float>(icon_itr->second->height)));
+        cursor_position.x += icon_itr->second->width;
```

Comment:
Is there a reason you're doing manual cursor management instead of letting imgui manage this for you?

Reply:
If I just let ImGui do this it'll stack the images vertically, and the recommended way according to Google/StackOverflow to put ImGui elements horizontally is by moving the cursor.

Reply:
Imgui will just do whatever you tell it to do with its layout functionality.  By default, if you call `Image` repeatedly, yes, it will stack them up one by one vertically on top of each other.

Putting aside the cursor management, do you really want to display things horizontally?  What if there's 100 icons?  Won't you run out of horizontal space?  I don't know anything about retroachievements but I'd assume you'd want some view [like this](https://www.reddit.com/media?url=https%3A%2F%2Fpreview.redd.it%2Fdlh1bj2s3vr91.png%3Fauto%3Dwebp%26s%3D8e3e6d39fbc0e742bcddd4b6808ad1d4a0967a37).

Reply:
I've been over the ImGui documentation and I never found anything about grouping things horizontally. What do you recommend?

Reply:
I guess it depends on what you want to do visually?  The screenshot is likely a table like construct with a max set of columns (and also probably a max set of rows)

Reply:
Now that I understand what is being done from the mockup I can give some more appropriate comments.  If you wanted to avoid manual cursor management, you'd use `ImGui::SameLine()` to keep the image from going to a new line.

This is subjective but I think it might also make sense to add a little padding between each image, I haven't used this but I believe you can  use [Dummy()](https://github.com/ocornut/imgui/blob/master/imgui.h#L485) to do that.